python_modules/dagster/dagster/_core/storage/upath_io_manager.py

```diff
@@ -0,0 +1,195 @@
+from __future__ import annotations
+
+import inspect
+from abc import abstractmethod
+from typing import Any, Dict, List, Union
+
+from upath import UPath
+
+from dagster import (
+    AssetKey,
+    IOManager,
+    InputContext,
+    MemoizableIOManager,
+    MetadataValue,
+    OutputContext,
+    TimeWindowPartitionsDefinition,
+)
+from dagster import _check as check
+
+
+class UPathIOManagerBase(MemoizableIOManager):
+    """
+    Abstract IOManager base class compatible with local and cloud storage via `fsspec` (using `universal-pathlib`).
+    What it handles for the use:
+     - working with any filesystem supported by `fsspec`
+     - handling loading multiple upstream asset partitions via PartitionMapping
+     (returns a dictionary with mappings from partition_keys to loaded objects)
+     - the `get_metadata` method can be customized to add additional metadata to the outputs
+     - the `allow_missing_partitions: bool` metadata value can control
+     either raising an error or skipping on missing partitions
+    """
+
+    extension: str = ""  # override in child class
+
+    @abstractmethod
+    def dump_to_path(self, context: OutputContext, obj: Any, path: UPath):
+        """
+        Child classes should override this method to write the object to the filesystem.
+
+        Args:
+            context:
+            obj:
+            path:
+
+        Returns:
+
+        """
+        ...
+
+    @abstractmethod
+    def load_from_path(self, context: InputContext, path: UPath) -> Any:
+        """
+        Child classes should override this method to load the object from the filesystem.
+
+        Args:
+            context:
+            path:
+
+        Returns:
+
+        """
+        ...
+
+    @staticmethod
+    def get_metadata(context: OutputContext, obj: Any) -> Dict[str, MetadataValue]:
```

Comment:
Nitpick: `Mapping` is preferred over `Dict`?

Also, any particular reason this needs to be static? It could potentially benefit from having access to `self` - e.g. to use the base path.